[MRG] Bugfix for not-yet-confirmed issue #12863: arpack returns singular values in ascending order, the opposite was supposed in sklearn

gykovacs opened the pull request:
Bugfix for not-yet-confirmed issue #12863: arpack returns singular values in ascending order, the opposite was supposed in sklearn

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes [MRG] not-yet-confirmed issue #12863 

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

In the sklearn code it was assumed that the scipy wrapper for arpack singular value decomposition returns singular values (and corresponding vectors) in descending order. This is not the case. The ARPACK documentation (https://www.caam.rice.edu/software/ARPACK/UG/node136.html#SECTION001210000000000000000
) clearly states that the singular values are returned in ascending order.

The bug manifests in differing results depending on the solver:

```
import numpy as np
from sklearn import cluster

A= np.array([[-2, -4, 2], [-2, 1, 2], [4, 2, 5]])

sc= cluster.bicluster.SpectralCoclustering(n_clusters= 2, 
                                           svd_method='randomized')

sc.fit(A)
print(sc.column_labels_)
```
gives
```
[0 0 1]
```
, but
```
sc= cluster.bicluster.SpectralCoclustering(n_clusters= 2, 
                                           svd_method='arpack')
sc.fit(A)
print(sc.column_labels_)
```
gives
```
[0 0 0]
```

This bugfix makes the arpack call return the same result as the randomized sv based solution.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

gykovacs committed:
arpack returns singular values in ascending order, the use of n_discard fixed accordingly

gykovacs committed:
arpack returns sv-s in ascending order, the use of n_discard fixed

gykovacs committed:
whitespace from blank lines removed

jnothman commented:
Please add a test.

gykovacs commented:
I'm working on it!

gykovacs committed:
test added to check if arpack based and randomized svd solvers give the same results in spectral coclustering

gykovacs committed:
trailing whitespace removed

gykovacs committed:
one more blank line added between tests

gykovacs committed:
another blank line added between tests

gykovacs commented:
I have added a test checking if the labels implied by the singular values computed by 'arpack' and 'randomized' methods are equal (the svs are not available in the BaseSpectral objects, so the direct comparison of the order of svs cannot be tested). 

I also tested the test, it passes with my updates to BaseSpectral._svd but fails without it, so the test matrix seems to be a good choice to drive the testing.

However, one CI job fails: the one related to OSX. Interestingly, another svd calculation fails, the numbers do not match the expectations in the 12th digit. In my best understanding this has nothing to do with my bugfix, this mismatch is based on an svd call made directly to scipy at line 163 of decomposition/truncated_svd.py. I'm wondering how to proceed. Shall I maybe decrease the required precision in this test to 11 digits?

jnothman commented:
I've not looked into the failure, but I think it *must* be due to your change, even if your change is right.

jnothman commented:
Yes, I think it should be okay to increase the tolerance.

jnothman commented:
(The rpca in that test code is a bit weird... both are using 'arpack', not 'randomized')

gykovacs committed:
test fixed by changing arpack parameter to randomized

gykovacs committed:
retry to see if the problem is numeric

gykovacs committed:
changes reverted to check if the test passes

gykovacs committed:
skipping the test added to see if arpack is likely to be stateful

gykovacs committed:
fix added again

jnothman commented:
Are you sure it's the change to the test, not the change to the implementation, that resulted in the failure?

jnothman commented:
I suspect that when that test comparing rpca to apca was adapted from test_pca, it was seen as inappropriate to compare apca and rpca, and somehow randomized was changed to arpack without thinking it through... not sure though.

jnothman commented:
So I don't think we should be testing randomized there, at least not in this PR.

gykovacs committed:
original PR + decrease of required precision in the failing test

gykovacs commented:
I did some further checks, let me summarize briefly:

1) originally, I added a fix to the bug, and all tests passed (https://travis-ci.org/scikit-learn/scikit-
learn/builds/474089316?utm_source=github_status&utm_medium=notification)

2) added a test to test the labels in coclustering implied by the ordering of svds, and another test failed (https://travis-ci.org/scikit-learn/scikit-learn/builds/482023389?utm_source=github_status&utm_medium=notification)

3) I put a SkipException into the test I added, and all tests passed (with my implementation changes in the package) (https://travis-ci.org/scikit-learn/scikit-learn/builds/482182493?utm_source=github_status&utm_medium=notification)

4) one more observation on why scipy arpack seems to be stateful: as you already mentioned, the failing test is not correct, as rpca uses arpack just like apca. Both of these objects are initialized by the same random state, thus, the objects are basically identical. Contrarily, the test fails because of a tiny difference in the 12th digit. In my interpretation there should be no way of having any difference between the results unless some execution path maintains some state, and my guess is that it comes from scipy, as it is the common point between the two tests.

Now, I have reverted the changes, the bugfix is in the code, my test is added, the precision of the improper (arpack vs arpack) test is decreased to 11 digits, and things seem to work fine, all tests are passing.

gykovacs committed:
empty commit to trigger new CI

gykovacs commented:
I did some further checks, let me summarize briefly:

1) originally, I added a fix to the bug, and all tests passed (https://travis-ci.org/scikit-learn/scikit-
learn/builds/474089316?utm_source=github_status&utm_medium=notification)

2) then, I added a test to test the labels in coclustering (which are implied by the ordering of svds), and another test failed (https://travis-ci.org/scikit-learn/scikit-learn/builds/482023389?utm_source=github_status&utm_medium=notification)

3) then, I put a SkipException into the test I added, and all tests passed (with my bugfix in the package), which suggests that not the bugfix, but the execution of the additional test causes the failure of another test (https://travis-ci.org/scikit-learn/scikit-learn/builds/482182493?utm_source=github_status&utm_medium=notification)

4) One more observation on why scipy arpack seems to be stateful: as you already mentioned, the failing test is not correct, as rpca uses arpack just like apca. Both of these objects are initialized by the same random state and parameters, thus, the objects are identical and should give identical results. Contrarily, the test fails because of a tiny difference in the 12th digit. In my interpretation there should be no way of having any difference between the results unless some execution path maintains some state or uncontrolled randomness, and my guess is that it comes from scipy, as it is the only common point between the two tests.

Now, I have the bugfix in the code, my test is added, the required precision of the failing (arpack vs arpack) test is decreased to 11 digits, and things seem to work fine, all tests are passing. Maybe it would worth reporting the failure of the arpack vs arpack comparison in the truncated svd test as a separate issue.

Any comments are welcome!

jnothman reviewed (commented):
Thanks. I've confirmed the test failing at master. 

Please add an entry to the change log at `doc/whats_new/v0.21.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

The arpack state issue might be worth raising separately.

gykovacs committed:
doc updated by bugfix documentation referring pull request 12898

gykovacs committed:
Merge branch 'master' into feature-arpack

gykovacs commented:
Done! Thank you for the guidance! I'll raise the arpack state issue and also try to take it soon.

cmarmo commented:
Hi @gykovacs , if you are still interested in working on that, do you mind fixing conflicts? Thanks for your work and your patience!

gykovacs commented:
Yeah, let me look into it!
